File: cmd/delete.go
package cmd

import (
	"bufio"
	"os"
	"path/filepath"
	"strings"

	"github.com/benoctopus/sesh/internal/config"
	"github.com/benoctopus/sesh/internal/display"
	"github.com/benoctopus/sesh/internal/git"
	"github.com/benoctopus/sesh/internal/models"
	"github.com/benoctopus/sesh/internal/project"
	"github.com/benoctopus/sesh/internal/session"
	"github.com/benoctopus/sesh/internal/state"
	"github.com/benoctopus/sesh/internal/workspace"
	"github.com/rotisserie/eris"
	"github.com/spf13/cobra"
)

var (
	deleteAll         bool
	deleteForce       bool
	deleteProjectName string
)

var deleteCmd = &cobra.Command{
	Use:     "delete [branch]",
	Aliases: []string{"del", "remove", "rm"},
	Short:   "Delete worktree, session, or entire project",
	Long: `Delete a worktree and its associated session, or delete an entire project.

By default, deletes the specified branch's worktree and session.
Use --all to delete the entire project including all worktrees.

The project is automatically detected from the current working directory,
or can be specified explicitly with the --project flag.

Examples:
  sesh delete feature-foo          # Delete feature-foo worktree/session
  sesh delete --all                # Delete entire project (requires confirmation)
  sesh delete --all --force        # Delete entire project without confirmation
  sesh delete --project myproject --all  # Delete specific project`,
	RunE: runDelete,
}

func init() {
	rootCmd.AddCommand(deleteCmd)
	deleteCmd.Flags().BoolVar(&deleteAll, "all", false, "Delete entire project")
	deleteCmd.Flags().BoolVarP(&deleteForce, "force", "f", false, "Skip confirmation prompt")
	deleteCmd.Flags().
		StringVarP(&deleteProjectName, "project", "p", "", "Specify project explicitly")
}

func runDelete(cmd *cobra.Command, args []string) error {
	disp := display.NewStderr()

	// Load configuration
	cfg, err := config.LoadConfig()
	if err != nil {
		return eris.Wrap(err, "failed to load configuration")
	}

	// Get current working directory
	cwd, err := os.Getwd()
	if err != nil {
		return eris.Wrap(err, "failed to get current working directory")
	}

	// Resolve project from filesystem state
	proj, err := project.ResolveProject(cfg.WorkspaceDir, deleteProjectName, cwd)
	if err != nil {
		return eris.Wrap(err, "failed to resolve project")
	}

	if deleteAll {
		return deleteProject(cfg, proj, disp)
	}

	// Delete specific branch
	if len(args) == 0 {
		return eris.New("branch name required (or use --all to delete entire project)")
	}

	branch := args[0]
	return deleteBranch(cfg, proj, branch, disp)
}

func deleteProject(cfg *config.Config, proj *models.Project, disp display.Printer) error {
	// Get all worktrees for this project
	worktrees, err := state.DiscoverWorktrees(proj)
	if err != nil {
		return eris.Wrap(err, "failed to discover worktrees")
	}

	if !deleteForce {
		// Ask for confirmation
		disp.Printf(
			"This will delete project '%s' with %d worktree(s) and all associated sessions.\n",
			proj.Name,
			len(worktrees),
		)
		disp.Printf("Project path: %s\n", proj.LocalPath)
		disp.Print("Are you sure? (yes/no): ")

		reader := bufio.NewReader(os.Stdin)
		response, err := reader.ReadString('\n')
		if err != nil {
			return eris.Wrap(err, "failed to read confirmation")
		}

		response = strings.TrimSpace(strings.ToLower(response))
		if response != "yes" && response != "y" {
			disp.Println("Deletion cancelled.")
			return nil
		}
	}

	// Initialize session manager
	sessionMgr, err := session.NewSessionManager(cfg.SessionBackend)
	if err != nil {
		return eris.Wrap(err, "failed to initialize session manager")
	}

	// Delete all sessions
	for _, wt := range worktrees {
		// Generate session name
		sessionName := workspace.GenerateSessionName(proj.Name, wt.Branch)

		// Kill session if it exists
		exists, err := sessionMgr.Exists(sessionName)
		if err != nil {
			disp.Printf("Warning: failed to check session existence: %v\n", err)
		} else if exists {
			disp.Printf("Killing %s session: %s\n", sessionMgr.Name(), sessionName)
			if err := sessionMgr.Delete(sessionName); err != nil {
				disp.Printf("Warning: failed to kill session: %v\n", err)
			}
		}

		// Remove worktree
		disp.Printf("Removing worktree: %s\n", wt.Path)
		if err := git.RemoveWorktree(proj.LocalPath, wt.Path); err != nil {
			disp.Printf("Warning: failed to remove worktree: %v\n", err)
		}
	}

	// Delete project directory (including bare repo)
	projectPath := filepath.Dir(proj.LocalPath)
	disp.Printf("Removing project directory: %s\n", projectPath)
	if err := os.RemoveAll(projectPath); err != nil {
		return eris.Wrap(err, "failed to remove project directory")
	}

	disp.Printf("\nSuccessfully deleted project: %s\n", proj.Name)
	return nil
}

func deleteBranch(cfg *config.Config, proj *models.Project, branch string, disp display.Printer) error {
	// Get worktree from filesystem state
	worktree, err := state.GetWorktree(proj, branch)
	if err != nil {
		return eris.Errorf("worktree for branch %s not found", branch)
	}

	// Check if this is the main worktree
	if worktree.IsMain {
		return eris.Errorf("cannot delete main worktree, use --all to delete the entire project")
	}

	if !deleteForce {
		// Ask for confirmation
		disp.Printf(
			"This will delete worktree for branch '%s' and its associated session.\n",
			branch,
		)
		disp.Printf("Worktree path: %s\n", worktree.Path)
		disp.Print("Are you sure? (yes/no): ")

		reader := bufio.NewReader(os.Stdin)
		response, err := reader.ReadString('\n')
		if err != nil {
			return eris.Wrap(err, "failed to read confirmation")
		}

		response = strings.TrimSpace(strings.ToLower(response))
		if response != "yes" && response != "y" {
			disp.Println("Deletion cancelled.")
			return nil
		}
	}

	// Initialize session manager
	sessionMgr, err := session.NewSessionManager(cfg.SessionBackend)
	if err != nil {
		return eris.Wrap(err, "failed to initialize session manager")
	}

	// Generate session name
	sessionName := workspace.GenerateSessionName(proj.Name, branch)

	// Kill session if it exists
	exists, err := sessionMgr.Exists(sessionName)
	if err != nil {
		disp.Printf("Warning: failed to check session existence: %v\n", err)
	} else if exists {
		disp.Printf("Killing %s session: %s\n", sessionMgr.Name(), sessionName)
		if err := sessionMgr.Delete(sessionName); err != nil {
			disp.Printf("Warning: failed to kill session: %v\n", err)
		}
	}

	// Remove worktree
	disp.Printf("Removing worktree: %s\n", worktree.Path)
	if err := git.RemoveWorktree(proj.LocalPath, worktree.Path); err != nil {
		return eris.Wrap(err, "failed to remove worktree")
	}

	disp.Printf("\nSuccessfully deleted worktree for branch: %s\n", branch)
	return nil
}

File: src/cli/delete.rs
use clap::Args;
use crate::error::Result;
use crate::config;
use crate::store::Store;
use crate::managers::{ProjectManager, WorktreeManager};
use std::io::{self, Write};
use tracing::info;

#[derive(Args)]
pub struct DeleteArgs {
    /// Project name to delete
    #[arg(short = 'p', long)]
    pub project: Option<String>,
    
    /// Worktree ID to delete
    #[arg(long)]
    pub worktree: Option<i64>,
    
    /// Skip confirmation prompt
    #[arg(short = 'f', long)]
    pub force: bool,
    
    /// Delete entire project (requires --project)
    #[arg(long)]
    pub all: bool,
}

fn confirm_deletion(message: &str, force: bool) -> Result<bool> {
    if force {
        return Ok(true);
    }
    
    // Check if we're in an interactive terminal
    let is_tty = std::io::IsTerminal::is_terminal(&std::io::stdin());
    if !is_tty {
        return Err(crate::error::Error::GitError {
            message: "--force flag required for deletion in noninteractive mode".to_string(),
        });
    }
    
    print!("{}", message);
    io::stdout().flush()?;
    
    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    
    let response = input.trim().to_lowercase();
    Ok(response == "yes" || response == "y")
}

pub async fn run(args: DeleteArgs) -> Result<()> {
    let config = config::load()?;
    let store = Store::open().await?;
    
    let project_manager = ProjectManager::new(store.clone(), config.clone());
    let worktree_manager = WorktreeManager::new(store.clone(), config);
    
    if let Some(project_name) = args.project {
        let project = project_manager.get_by_name(&project_name).await?;
        
        if args.all {
            // Delete entire project
            let worktrees = crate::store::queries::list_worktrees(store.pool(), project.id).await?;
            let message = format!(
                "This will delete project '{}' with {} worktree(s) and all associated sessions.\nProject path: {}\nAre you sure? (yes/no): ",
                project_name,
                worktrees.len(),
                project.clone_path.display()
            );
            
            if !confirm_deletion(&message, args.force)? {
                println!("Deletion cancelled.");
                return Ok(());
            }
        }
        
        project_manager.delete(project.id).await?;
        info!(project = %project_name, "Project deleted");
        println!("Deleted project: {}", project_name);
    } else if let Some(worktree_id) = args.worktree {
        let worktree = crate::store::queries::get_worktree(store.pool(), worktree_id).await?;
        let message = format!(
            "This will delete worktree for branch '{}' and its associated session.\nWorktree path: {}\nAre you sure? (yes/no): ",
            worktree.branch,
            worktree.path.display()
        );
        
        if !confirm_deletion(&message, args.force)? {
            println!("Deletion cancelled.");
            return Ok(());
        }
        
        worktree_manager.delete(worktree_id).await?;
        info!(worktree_id, "Worktree deleted");
        println!("Deleted worktree: {}", worktree_id);
    } else {
        eprintln!("Must specify either --project or --worktree");
    }
    
    Ok(())
}

